server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,187 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.get;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.engine.InternalEngine;
+import org.elasticsearch.index.get.GetResult;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+// TODO(ES-5727): add a retry mechanism to TransportGetFromTranslogAction
+public class TransportGetFromTranslogAction extends HandledTransportAction<
+    TransportGetFromTranslogAction.Request,
+    TransportGetFromTranslogAction.Response> {
+
+    public static final String NAME = "internal:data/read/get_from_translog";
+    public static final Logger logger = LogManager.getLogger(TransportGetFromTranslogAction.class);
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, Request::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final GetRequest getRequest = request.getRequest();
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        assert indexShard.routingEntry().isPromotableToPrimary();
+        assert getRequest.realtime();
+        Engine engine = indexShard.getEngineOrNull();
+        if (engine == null) {
+            listener.onFailure(new AlreadyClosedException("engine closed"));
+            return;
+        }
```

Comment:
I wonder if this is really needed because `ShardGetService.getFromTranslog()` uses  `getEngine()` at some point and that will throw in case of a closed engine. I think we can do this only if the `result` of the `getFromTranslog` is null?